native/exe/ServiceControlLaunch/ServiceControlLaunch.cpp

```diff
@@ -126,44 +125,71 @@ int APIENTRY WinMain(HINSTANCE hInstance,
 	DWORD dwRead = 0;
 	DWORD hsize = sizeof(dwRead);
 	hsize = sizeof(currVer);
-	if (RegOpenKeyEx(rootKey, "Software\\JavaSoft\\Java Runtime Environment", 0, KEY_QUERY_VALUE, &myKey) != ERROR_SUCCESS)
+	HMODULE exeMod = GetModuleHandle(NULL);
+	char appPath[512];
+	GetModuleFileName(exeMod, appPath, sizeof(appPath));
+	size_t appLen = strlen(appPath);
+	if (appLen > 0)  // Shouldn't be 0 as the following would be an infinite loop
 	{
-		JVM_MISSING;
-		return FALSE;
+		for (size_t i = appLen - 1; i > 0; i--)
+		{
+			if (appPath[i] == '\\')
+			{
+				appPath[i + 1] = 0;
+				break;
+			}
+		}
 	}
-	
-	if (RegQueryValueEx(myKey, "CurrentVersion", 0, &readType, (LPBYTE)currVer, &hsize) != ERROR_SUCCESS)
+
+	// See if we've got a JVM in our own directory to load
+	char includedJRE[1024];
+	strcpy_s(includedJRE, sizeof(includedJRE), appPath);
+	strcat_s(includedJRE, sizeof(includedJRE), "jre\\bin\\client\\jvm.dll");
+	HMODULE hLib = LoadLibrary(includedJRE);
+
+	if (hLib == NULL)
 	{
+		// Failed to find JRE in SageTV directory, load the JVM from the registry instead, by using the Windows Registry to locate the current version to use.
+
+		if (RegOpenKeyEx(rootKey, "Software\\JavaSoft\\Java Runtime Environment", 0, KEY_QUERY_VALUE, &myKey) != ERROR_SUCCESS)
+		{
+			JVM_MISSING;
+			return FALSE;
+		}
+
+		if (RegQueryValueEx(myKey, "CurrentVersion", 0, &readType, (LPBYTE)currVer, &hsize) != ERROR_SUCCESS)
+		{
+			RegCloseKey(myKey);
+			JVM_MISSING;
+			return FALSE;
+		}
 		RegCloseKey(myKey);
-		JVM_MISSING;
-		return FALSE;
-	}
-	RegCloseKey(myKey);
-	char pathKey[1024];
-	strcpy(pathKey, "Software\\JavaSoft\\Java Runtime Environment\\");
-	strcat(pathKey, currVer);
-	char jvmPath[1024];
-	if (RegOpenKeyEx(rootKey, pathKey, 0, KEY_QUERY_VALUE, &myKey) != ERROR_SUCCESS)
-	{
-		JVM_MISSING;
-		return FALSE;
-	}
-	hsize = sizeof(jvmPath);
-	if (RegQueryValueEx(myKey, "RuntimeLib", 0, &readType, (LPBYTE)jvmPath, &hsize) != ERROR_SUCCESS)
-	{
+		char pathKey[1024];
+		strcpy_s(pathKey, sizeof(pathKey), "Software\\JavaSoft\\Java Runtime Environment\\");
+		strcat_s(pathKey, sizeof(pathKey), currVer);
+		char jvmPath[1024];

```

Comment:
Use _MAX_PATH again here for jvmPath....registry values are different, so don't use _MAX_PATH for pathKey (leave it alone).

Reply:
Done